fix(Msal-Angular): Dynamic interceptor authority

svrooij opened the pull request:
Helps supporting true multi tenant applications, by allowing
overriding the authority at runtime.

Fixed #3223

jasonnutter commented:
@svrooij Thanks for the PR, we'll take a look and let you know if we have feedback.

svrooij commented:
Thanks @jasonnutter I was already discussing this with @tnorling in #3223

But since I missed the tests part and the coverage went down by 7% I'm going to add some tests later this evening (UTC+1 here).

svrooij commented:
@jasonnutter according to the [docs](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/contributing.md#creating-a-pr) I also have to do some **Beachball** stuff, is that something I should do or is that something you as a code owner should do? Since that probably will end up in the release notes somewhere.

jasonnutter commented:
> @jasonnutter according to the [docs](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/contributing.md#creating-a-pr) I also have to do some **Beachball** stuff, is that something I should do or is that something you as a code owner should do? Since that probably will end up in the release notes somewhere.

This is something you should do. In your branch, run `npm install` in the root of repo, and then run `npm run beachball:change` and follow the prompts, and then push the updated branch.

svrooij commented:
@jasonnutter already did that. Was a small change ☝️

tnorling commented:
@svrooij I think I would rather handle this in `msal-browser`, if possible, as this will likely affect more users than just Angular users. But I'm not sure what that would look like, we may need to add an API to `PublicClientApplication` to make changes to the base config after instantiation.

svrooij commented:
I partially agree with you @tnorling, in that this will most likely affect MSAL-browser users as well.

Since this is my first contribution I did not want to go ahead and change the public api of the most important part of this entire library. Also changing something big like that would require some discussions with the code owners I guess?

I asked myself, does this change break stuff for existing users and in what way can we make it that it will fit multiple use cases. If this way is liked by many users, we could also implement something similar for MSAL-browser.

jasonnutter commented on lib/msal-angular/docs/multi-tenant.md:
```diff
@@ -45,3 +45,37 @@ export class AppComponent implements OnInit {
+    authRequest: {
+      authority: 'https://login.microsoftonline.com/organizations',
+    },
+    dynamicAuthority: (account: AccountInfo) => `https://login.microsoftonline.com/${account.tenantId ?? 'organizations'}` 
```
Instead of just changing the authority, we can maybe consider being able to change the entire `authRequest`? This would mean changing `authRequest` to support an object literal or a function (that takes in the account, like your PR shows) that returns an object literal. That would help address other scenarios for other properties on the `authRequest`.

jasonnutter replied:
Also, if we can make a similar change for MSAL Guard, I think that would beneficial.

svrooij replied:
I did consider this, but came to the conclusion that you cannot change other properties of the request. The scope is already passed in and changing the redirect uri at runtime is weird since it requires registration at the azure portal.

Also you cannot change any properties on the guard, because either you have an account and the guard just says "ok go ahead" or you don't have an account and then there is no account to change it dynamically.

jasonnutter replied:
Any of the properties for the declared types (`Omit<PopupRequest, "scopes"> | Omit<RedirectRequest, "scopes"> | Omit<SilentRequest, "scopes">;`) are changable.

jasonnutter replied:
(We may consider adding `account` to teh omit list, cc: @jo-arroyo)

jasonnutter replied:
For example:

```js
let authRequest;
if (typeof this.msalInterceptorConfig.authRequest === "function") {
  authRequest = this.msalInterceptorConfig.authRequest(account);
} else {
  authRequest = this.msalInterceptorConfig.authRequest;
}
```

svrooij replied:
I know those properties can be changed because of the type definition, but I was wondering what properties would be reasonable to change.

And omitting the account might be a good idea 😉

jasonnutter replied:
> I know those properties can be changed because of the type definition, but I was wondering what properties would be reasonable to change.

We designed it to allow any of the properties (other than `scopes`) to be set via `authRequest`, and I don't see any reason not to allow of those same properties to be set via a dynamic `authRequest`.

jasonnutter replied:
In fact, a few of them (e.g. correlation id) may only be useful if they can be set dynamically.

jasonnutter replied:
> And omitting the account might be a good idea

I would actually wait on that. Right now we support the active account or the first account from `getAllAccounts` array, but we may apps to have different logic.

svrooij replied:
So the logic can be something like `dynamicAuthRequest: (originalAuthRequest: ..., httpReq: HttpEvent<any>): AuthRequest`

Then you can support overriding all properties you want. And you're right it does make sense to set the CorrelationId dynamically.

jasonnutter replied:
I would personally prefer the `authRequest` property support two types (typed object or function that returns a typed object), instead of having two separate fields, as I think that will be simpler for us and applications.

@jo-arroyo @tnorling Thoughts?

jasonnutter replied:
And as mentioned, I think we should have this change for both the guard and interceptor, given they have the same considerations around building the auth request.

svrooij replied:
@jasonnutter it also makes sense to still have defaults defined. By having two types on the authRequest configuration you cannot use both at the same time. And have to use dynamic for everything.

jasonnutter replied:
> By having two types on the authRequest configuration you cannot use both at the same time.

True, but if you need to set static and dynamic fields, both can be done in the dynamic version, e.g.

```js
authRequest: (httpReq, msalInstance) => {
  return {
     correlation: dyanmicCorrelationId,
     forceRefesh: true
  };
};
```

jasonnutter replied:
> it also makes sense to still have defaults defined

Can you please clarify what you mean? The scope and account are the only properties that we set, we don't provide any additional defaults (aside from the defaults set internally by MSAL for acquireTokenSilent, etc).

tnorling replied:
I like the idea of having one `authRequest` property that can either be static or a function. I can forsee lots of confusion about which is used, which should be used, what the difference is, etc. if we keep them separate.

svrooij replied:
I changed it to be one property that is either a function or a static. And added the same functionality for the MsalGuard

jasonnutter commented:
> @svrooij I think I would rather handle this in `msal-browser`, if possible, as this will likely affect more users than just Angular users. But I'm not sure what that would look like, we may need to add an API to `PublicClientApplication` to make changes to the base config after instantiation.

For the purposes of MSAL Angular, I think we can limit the scope of this change to just changing the properties passed in teh requests made in teh interceptor and guard (the `authRequest` property), and not the base configuration.

svrooij commented:
@jasonnutter what properties are considered changeable? I'm not very known to the library just yet.

jasonnutter commented on lib/msal-angular/src/msal.interceptor.config.ts:
```diff
@@ -3,10 +3,13 @@
     interactionType: InteractionType.Popup | InteractionType.Redirect;
     protectedResourceMap: Map<string, Array<string>>;
-    authRequest?: Omit<PopupRequest, "scopes"> | Omit<RedirectRequest, "scopes"> | Omit<SilentRequest, "scopes">;
+    authRequest?: MsalInterceptorAuthRequest | ((originalAuthRequest: MsalInterceptorAuthRequest, req: HttpRequest<unknown>) => MsalInterceptorAuthRequest);
```
```suggestion
    authRequest?: MsalInterceptorAuthRequest | ((msalService: MsalService, req: HttpRequest<unknown>) => MsalInterceptorAuthRequest);
```

svrooij replied:
The MSAL service doesn't provide exposed configuration. And the select account functionality is already build into the interceptor. 
This way it seems more like a way to override then to replace (what you are proposing).

svrooij replied:
Currently the code in the interceptor is configured like this:
```js
const authRequest = typeof this.msalInterceptorConfig.authRequest === "function"
            ? this.msalInterceptorConfig.authRequest({ account: account }, req)
            : { ...this.msalInterceptorConfig.authRequest, account };
```

And just above that part is the logic to get to correct account.

That means the authRequest can be overriden like:
```js
authRequest: (authRequest, httpReq) => { 
  return { ...authRequest, authority: 'something here' };
}
```

And you don't have to copy the select the correct account logic. How about I add the MsalService as well, that way you can do both. So get the account in a none default way and just use the default way and override only the stuff you want overridden?

jasonnutter commented on lib/msal-angular/docs/multi-tenant.md:
```diff
@@ -45,3 +45,42 @@ export class AppComponent implements OnInit {
+    authRequest: {
+      authority: 'https://login.microsoftonline.com/organizations',
+    },
+    dynamicAuthRequest: (authRequest, httpReq) => {
```
```suggestion
    authRequest: (msalService, httpReq) => {
```

jasonnutter commented on lib/msal-angular/docs/multi-tenant.md:
```diff
@@ -45,3 +45,42 @@ export class AppComponent implements OnInit {
+    dynamicAuthRequest: (authRequest, httpReq) => {
+      return {
+        ...authRequest,
+        authority: `https://login.microsoftonline.com/${authRequest.account?.tenantId ?? 'organizations'}`
```
Instead of pull this off of the authRequest, you should use the get account APIs on the MSAL instance (either getActiveAccount, getAllAccounts[0], etc) depending on if your app sets the active account.

jasonnutter commented on lib/msal-angular/src/msal.guard.config.ts:
```diff
@@ -4,9 +4,12 @@
 export type MsalGuardConfiguration = {
     interactionType: InteractionType.Popup | InteractionType.Redirect;
-    authRequest?: Partial<PopupRequest> | Partial<Omit<RedirectRequest, "redirectStartPage">>;
+    authRequest?: MsalGuardAuthRequest | ((config: MsalGuardConfiguration, authService: MsalService) => MsalGuardAuthRequest);
```
We would like to provide the authSerivce and `state`

```suggestion
    authRequest?: MsalGuardAuthRequest | ((msalService: MsalService, state: RouterStateSnapshot) => MsalGuardAuthRequest);
```

svrooij replied:
This makes perfect sense, applied

svrooij commented:
And also fixed some coverage 🎉 

![image](https://user-images.githubusercontent.com/1292510/112466190-db932900-8d65-11eb-8d7a-78b5b21357d9.png)

svrooij commented:
@jasonnutter did you see my changes?

jasonnutter commented:
@svrooij Thanks, been busy but will look now.

svrooij commented:
Do you guys normally merge PRs or are they squashed? Since here are so many commits, maybe nicer if they are just merged as one commit.

jasonnutter commented on lib/msal-angular/src/msal.interceptor.config.ts:
```diff
@@ -3,10 +3,14 @@
     interactionType: InteractionType.Popup | InteractionType.Redirect;
     protectedResourceMap: Map<string, Array<string>>;
-    authRequest?: Omit<PopupRequest, "scopes"> | Omit<RedirectRequest, "scopes"> | Omit<SilentRequest, "scopes">;
+    authRequest?: MsalInterceptorAuthRequest | ((originalAuthRequest: MsalInterceptorAuthRequest, msalService: MsalService, req: HttpRequest<unknown>) => MsalInterceptorAuthRequest);
```
I guess I can see why it would be useful, but I would prefer it is put at the end:

```suggestion
    authRequest?: MsalInterceptorAuthRequest | ((msalService: MsalService, req: HttpRequest<unknown>, originalAuthRequest: MsalInterceptorAuthRequest) => MsalInterceptorAuthRequest);
```

svrooij replied:
My original comment got lost somewhere, but this makes it possible to use the account selection logic from the interceptor and just override the properties you want.

If we remove the original auth request the developer is required to do the account selection, and this becomes more an provide your own implementation instead of change some properties to your liking.

By providing both the MSAL service and the computed authentication request the developer has a choice to either set own implementation or just override for instance the authority based on the selected account

svrooij replied:
I’ll move it to the back and update the code calling this, tomorrow from behind my Pc cannot do that on my tablet while being on the sofa.

jasonnutter commented:
> Do you guys normally merge PRs or are they squashed? Since here are so many commits, maybe nicer if they are just merged as one commit.

Normally just merge, but can squash this one since it has so many commits. Not a big deal for us, frankly.

svrooij commented:
@jasonnutter I've tried to keep up with the changes on the dev branch, but after the last 2 merges, it's again out-of-date. Does that matter?

jasonnutter commented:
> @jasonnutter I've tried to keep up with the changes on the dev branch, but after the last 2 merges, it's again out-of-date. Does that matter? Invited you to my fork, so you should be able to fix merge stuff also.

Don't worry about that, we can handle keeping that up to date.

svrooij commented:
@jasonnutter Can I do something to move this PR forward? Or did I miss one of your comments? If you guys are busy that is also fine, I just have no clue how long a regular pr would take for this library.

jasonnutter commented on lib/msal-angular/src/msal.guard.spec.ts:
```diff
@@ -13,21 +13,26 @@ let routeStateMock: any = { snapshot: {}, url: '/' };
       clientId: '6226576d-37e9-49eb-b201-ec1eeb0029b6',
       redirectUri: 'http://localhost:4200'
-    }
+    },
+    //system: browserSystemOptions
```
Please remove

jasonnutter replied:
```suggestion
    }
```

jasonnutter commented on lib/msal-angular/src/msal.guard.config.ts:
```diff
@@ -3,10 +3,16 @@
-    authRequest?: Partial<PopupRequest> | Partial<Omit<RedirectRequest, "redirectStartPage">>;
+    authRequest?: MsalGuardAuthRequest | ((authService: MsalService, state: RouterStateSnapshot) => MsalGuardAuthRequest);
     loginFailedRoute?: string;
+    canActivate?: (msalService: MsalService, state?: RouterStateSnapshot) => Observable<boolean> | boolean;
```
I'm not sure I understand why this is necessary? It seems like this would allow someone to implement their own guard logic, in which case, we would rather an app implement a custom guard, instead of using the MSAL Guard (which is intended to be generic, and will not work for all use cases).

jasonnutter replied:
Furthermore, this feels like scope creep for this PR. I would prefer it is scoped to just the dynamic auth request, and have further enhacements be discussed/implemented separately.

jasonnutter commented on lib/msal-angular/docs/multi-tenant.md:
````diff
@@ -45,3 +45,39 @@ export class AppComponent implements OnInit {
   }
 ```
+
+## Multi-tenant access tokens
````
I think it would be good to include two sub sections here, one for the guard, and one for the interceptor.

jasonnutter commented:
> @jasonnutter Can I do something to move this PR forward? Or did I miss one of your comments? If you guys are busy that is also fine, I just have no clue how long a regular pr would take for this library.

Sorry, been busy. Left a few comments.

svrooij commented:
@jasonnutter I removed the feature creep and created a new issue for it #3406 I also changed the documentation and removed some comments in the tests that got left over from trying to debug some issue I was having.

jasonnutter commented:
> @jasonnutter I removed the feature creep and created a new issue for it #3406 I also changed the documentation and removed some comments in the tests that got left over from trying to debug some issue I was having.

Great, thanks! Will take a look.

jasonnutter reviewed (approved):
Looks good, thanks for addressing feedback! @jo-arroyo Can you please take a look when you get a chance?

jo-arroyo commented on lib/msal-angular/docs/multi-tenant.md:
```diff
@@ -45,3 +45,46 @@ export class AppComponent implements OnInit {
+
+## Dynamic auth request
+
+By default the MsalGuard and the MsalInterceptor use the static properties you set in the config. Both can be configured with a method for the `authRequest`. This allows to dynamically change the parameters used for authentication.
```
```suggestion
By default the MsalGuard and the MsalInterceptor use the static properties set in the config. Both can also be configured with a method for the `authRequest`, allowing the parameters used for authentication to be changed dynamically.
```

jo-arroyo commented on lib/msal-angular/docs/multi-tenant.md:
```diff
@@ -45,3 +45,46 @@ export class AppComponent implements OnInit {
+
+### MsalInterceptor - dynamic auth request (multi tenant tokens)
+
+If you're using `organizations` or `common` all the tokens will be requested for the users' home tenant, this might not result in the desired effect. If you have this application where you're invited as a guest the tokens will be from the wrong authority.
```
```suggestion
If `organizations` or `common` is used as the tenant, all tokens will be requested for the users' home tenant. However, this may not be the desired outcome. If a user is invited as a guest, the tokens may be from the wrong authority.
```

jo-arroyo commented on lib/msal-angular/docs/multi-tenant.md:
```diff
@@ -45,3 +45,46 @@ export class AppComponent implements OnInit {
+
+If you're using `organizations` or `common` all the tokens will be requested for the users' home tenant, this might not result in the desired effect. If you have this application where you're invited as a guest the tokens will be from the wrong authority.
+
+Setting the `authRequest` in the **MsalInterceptorConfig** to a method allows you to dynamically change the auth request, and for instance set the authority based on the home tenant of the account (when using guest users).
```
```suggestion
Setting the `authRequest` in the **MsalInterceptorConfig** to a method allows you to dynamically change the auth request. For instance, you may set the authority based on the home tenant of the account when using guest users.
```

jo-arroyo commented on lib/msal-angular/docs/multi-tenant.md:
```diff
@@ -45,3 +45,46 @@ export class AppComponent implements OnInit {
+If you're using `organizations` or `common` all the tokens will be requested for the users' home tenant, this might not result in the desired effect. If you have this application where you're invited as a guest the tokens will be from the wrong authority.
+
+Setting the `authRequest` in the **MsalInterceptorConfig** to a method allows you to dynamically change the auth request, and for instance set the authority based on the home tenant of the account (when using guest users).
+You can change all properties but be sure to always extend the `originalAuthRequest` like you see below.
```
```suggestion
Properties on `authRequest` may be changed, but should always extend the `originalAuthRequest` like below:
```

jo-arroyo commented on lib/msal-angular/docs/v2-docs/configuration.md:
```diff
@@ -270,3 +270,23 @@ fetch('/assets/configuration.json')
+
+## MsalGuard - dynamic auth request
+
+The **MsalGuard** also allows you to dynamically change the **authRequest** at runtime. This allow you to pick a different authority for some route, or to dynamically add scopes based on the **RouterStateSnapshot**.
```
```suggestion
The **MsalGuard** also allows you to dynamically change the **authRequest** at runtime. This allow you to pick a different authority for a route, or to dynamically add scopes based on the **RouterStateSnapshot**.
```

jo-arroyo commented on lib/msal-angular/src/msal.interceptor.spec.ts:
```diff
@@ -349,7 +353,60 @@ describe('MsalInterceptor', () => {
+    initializeMsal();
+    spyOn(PublicClientApplication.prototype, "getAllAccounts").and.returnValue([sampleAccountInfo]);
+    spyOn(PublicClientApplication.prototype, "acquireTokenSilent").and.callFake((silentRequest: SilentRequest) => new Promise((resolve) => {
+      // expect(silentRequest.authority).toBeUndefined();
```
Is this still needed here?

jo-arroyo commented on lib/msal-angular/src/msal.redirect.component.spec.ts:
```diff
@@ -0,0 +1,60 @@
+      MsalModule.forRoot(MSALInstanceFactory(), null, null)
+    ],
+    providers: [
+
```
```suggestion
```

jo-arroyo commented on lib/msal-angular/src/msal.redirect.component.spec.ts:
```diff
@@ -0,0 +1,60 @@
+  TestBed.configureTestingModule({
+    declarations: [MsalRedirectComponent],
+    imports: [
+
```
```suggestion
```

jo-arroyo commented on lib/msal-angular/src/msal.redirect.component.spec.ts:
```diff
@@ -0,0 +1,60 @@
+      clientId: '6226576d-37e9-49eb-b201-ec1eeb0029b6',
+      redirectUri: 'http://localhost:4200'
+    },
+    //system: browserSystemOptions
```
Is this still needed here?

jo-arroyo reviewed (commented):
Thanks for your work. I reviewed and have a few comments.

svrooij commented:
> @svrooij Thanks for your work. I reviewed and have a few comments.

@jo-arroyo Thank you for the feedback.
